Account/FxAClient10.swift

```diff
@@ -362,4 +411,134 @@ public class FxAClient10 {
+                        deferred.fill(Maybe(failure: FxAClientError.Local(error)))
+                        return
+                    }
+
+                    if let data = result.value {
+                        let json = JSON(data)
+                        if let remoteError = FxAClient10.remoteErrorFromJSON(json, statusCode: response!.statusCode) {
+                            deferred.fill(Maybe(failure: FxAClientError.Remote(remoteError)))
+                            return
+                        }
+
+                        if let response = FxAClient10.devicesResponseFromJSON(json) {
+                            deferred.fill(Maybe(success: response))
+                            return
+                        }
+                    }
+
+                    deferred.fill(Maybe(failure: FxAClientError.Local(FxAClientUnknownError)))
+                }
+        }
+        return deferred
+    }
+
+    public func registerOrUpdateDevice(sessionToken: NSData, device: FxADevice) -> Deferred<Maybe<FxADevice>> {
+        let deferred = Deferred<Maybe<FxADevice>>()
+        let salt: NSData = NSData()
+        let contextInfo: NSData = FxAClient10.KW("sessionToken")
+        let bytes = sessionToken.deriveHKDFSHA256KeyWithSalt(salt, contextInfo: contextInfo, length: UInt(2 * KeyLength))
+        let tokenId = bytes.subdataWithRange(NSMakeRange(0 * KeyLength, KeyLength))
+        let reqHMACKey = bytes.subdataWithRange(NSMakeRange(1 * KeyLength, KeyLength))
+        let hawkHelper = HawkHelper(id: tokenId.hexEncodedString, key: reqHMACKey)
+
+        let URL = self.URL.URLByAppendingPathComponent("/account/device")
+        let mutableURLRequest = NSMutableURLRequest(URL: URL)
+        mutableURLRequest.HTTPMethod = Method.POST.rawValue
+
+        mutableURLRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
+        mutableURLRequest.HTTPBody = device.toJSON().toString(false).utf8EncodedData
+
+        let hawkValue = hawkHelper.getAuthorizationValueFor(mutableURLRequest)
+        mutableURLRequest.setValue(hawkValue, forHTTPHeaderField: "Authorization")
+
+        alamofire.request(mutableURLRequest)
+            .validate(contentType: ["application/json"])
+            .responseJSON { (request, response, result) in
+                withExtendedLifetime(self.alamofire) {
+                    if let error = result.error as? NSError {
+                        deferred.fill(Maybe(failure: FxAClientError.Local(error)))
+                        return
+                    }
+
+                    if let data = result.value {
+                        let json = JSON(data)
+                        if let remoteError = FxAClient10.remoteErrorFromJSON(json, statusCode: response!.statusCode) {
+                            deferred.fill(Maybe(failure: FxAClientError.Remote(remoteError)))
+                            return
+                        }
+
+                        if let responseDevice = FxADevice.fromJSON(json) {
```

Comment:
Nit: Similar to the above comment, the way we handle the error in all three of these methods is the same. There's this neat pattern Richard did in the Sync code [1] that we might be able to use here to DRY this up. By passing the deferred through an `errorWrap` function and passing in the endpoint-specific code for formulating the response as a closure I think we can reduce the duplication.

[1] https://github.com/mozilla/firefox-ios/blob/master/Sync/StorageClient.swift#L331